grpcio-tools protobuf dependency issue

FamousPond opened the pull request:
As stated in issue #919, FRR zebra was hanging when Quagga was present on my machine. 

I was also unable to receive more than 64KB of data from commands through `/core/daemon/core/api/grpc/client.py`. When running a cat command through the `node_command` function on a file larger than 64KB it would just hang. 

Neither the GRPC client nor server had its max message size changed and so should have been able to send up to 4MB as that is the default.  

I was able to solve this issue by updating the protobuf version in the grpcio-tools dependencies from `>=5.26.1,<6.0dev` to `>=5.29.3,<6.0dev`.

This ensures that the same version of protobuf is used system wide.  

This issue was occurring on Ubuntu 22.04 and 24.04.

bharnden commented:
`This ensures that the same version of protobuf is used system wide.`

What do you mean by this?

FamousPond commented:
The issue is not only FRR hanging #919 but also when using GRPC to extract data from the node I was limited to 65KB instead of the expected 4MB. 

I think the issue is due to a protobuf misalignment somewhere. The only place that didn't define protobuf as version 5.29.3 was the protobuf dependency in grpcio-tools. 

Updating the poetry.lock file as I did fixed the issue for me. 

Yes it resolved #919. 


When I spin up a core instance then use the `node_command` function in the file `/core/daemon/core/api/grpc/client.py` to run the command `cat <file name>` on a file larger then 65,536 bytes It just hangs as I get no stdout or stderr output. 


Sorry it is a confusing bug and I struggle to explain it properly.

bharnden commented:
Looking at this further, it might have more to do with the build phase which has a separate path of dependencies and not within this lockfile at the location you changed. Since the lockfile itself has a lock for the specific version of protobuf it uses.

https://github.com/coreemu/core/blob/2682ddfe3e0536f7427c880fc17ec1bd1ed82877/daemon/poetry.lock#L421-L427

Maybe added a hard requirement for the version used when building would help, which is in `setup.sh`, but the version obtained here within the virtualenv would not be impacted.

FamousPond commented:
I believe you are correct. 
I reverted the poetry.lock file back and added in protobuf to the setup.sh file and It fixed the previous issue. 

I am wondering if, due to the order of the dependencies it is installing grpcio-tools with a different version of protobuf, then installing the expected version of protobuf just afterwards. 
Either way by putting protobuf into the setup.sh, it will subsequently use that version as it is cached.

bharnden commented:
I will look to add the same specific version request for the build dependency, thanks
